test/e2e/cluster_upgrade_test.go

```diff
@@ -69,9 +83,115 @@ var _ = Describe("When upgrading a workload cluster using ClusterClass with a HA
+					// * workload cluster nodes
+					// * kube-proxy pods
+					var upgraded int64
+					deletingMachinesWithPreDrainHook := []clusterv1.Machine{}
+					for _, m := range machines {
+						if *m.Spec.Version == cluster.Spec.Topology.Version && conditions.IsTrue(&m, clusterv1.MachineNodeHealthyCondition) {
+							upgraded++
+						}
+						if !m.DeletionTimestamp.IsZero() && m.Annotations[preDrainHook] == "true" {
+							deletingMachinesWithPreDrainHook = append(deletingMachinesWithPreDrainHook, m)
+						}
+					}
+
+					wlClient := managementClusterProxy.GetWorkloadCluster(ctx, workloadClusterNamespace, workloadClusterName).GetClient()
+					nodes := corev1.NodeList{}
+					if err := wlClient.List(ctx, &nodes); err != nil {
+						return 0, errors.Wrapf(err, "failed to list nodes in workload cluster")
+					}
+
+					kubeProxyPods := corev1.PodList{}
+					if err := wlClient.List(ctx, &kubeProxyPods, client.InNamespace(metav1.NamespaceSystem), client.MatchingLabels{"k8s-app": "kube-proxy"}); err != nil {
+						return 0, errors.Wrapf(err, "failed to list kube-proxy pods in workload cluster")
+					}
+
+					errList := []error{}
+
+					// Check all nodes to be Ready.
+					for _, node := range nodes.Items {
+						for _, condition := range node.Status.Conditions {
+							if condition.Type == corev1.NodeReady && condition.Status != corev1.ConditionTrue {
+								errList = append(errList, errors.Errorf("Expected the Ready condition for Node %s to be true but got %s instead: %s", node.GetName(), condition.Status, condition.Message))
+							}
+						}
+					}
+
+					// Check if the expected number of kube-proxy pods exist and all of them are healthy for all existing Nodes of the Cluster.
+					if len(nodes.Items) != len(kubeProxyPods.Items) {
+						errList = append(errList, errors.Errorf("exected %d kube-proxy pods to exist, got %d", len(nodes.Items), len(kubeProxyPods.Items)))
+					}
+					for _, pod := range kubeProxyPods.Items {
+						for _, condition := range pod.Status.Conditions {
+							if condition.Type == corev1.PodReady && condition.Status != corev1.ConditionTrue {
+								errList = append(errList, errors.Errorf("Expected the Ready condition for Pod %s to be true but got %s instead: %s", pod.GetName(), condition.Status, condition.Message))
+							}
+						}
+					}
+
+					if err := kerrors.NewAggregate(errList); err != nil {
+						return 0, errors.Wrap(err, "blocking upgrade because cluster is not stable")
+					}
+
+					// At this stage all current machines are considered ok, so remove the pre-drain webhook from a CP to unblock the next step of the upgrade.
+					if len(deletingMachinesWithPreDrainHook) > 0 {
+						if len(deletingMachinesWithPreDrainHook) > 1 {
+							return 0, errors.Errorf("expected a maximum of 1 control-plane machines to be in deleting and having the pre-drain hook but got %d", len(deletingMachinesWithPreDrainHook))
+						}
+
+						// Removing pre-drain hook from machine.
```

Comment:
Should we only do this if there are no Machines with deletionTimestamp and without pre-drain hook?

(just to ensure we're not hitting strange edge cases where we are removing pre-drain hooks from multiple machines)